docs: add platforms reference and how-to (#2080)
canonical/charmcraft: docs/howto/build-guides/select-platforms.rst

.. _select-platforms:

Select charm platforms
======================

> The ``platforms`` and ``base`` keys in a charm's :ref:`charmcraft-yaml-file` selects

Reviewer: ```suggestion
The ``platforms`` and ``base`` keys in a charm's :ref:`charmcraft-yaml-file` select
```

> both the operating system releases and silicon architectures on which the charm

Reviewer: ```suggestion
both the OS releases and silicon architectures on which the charm
```

can build and run.

See the :doc:`Platforms reference </reference/platforms>` for more information.

> Select an operating system

Reviewer: ```suggestion
Select an OS
```

For consistency.

--------------------------

The :ref:`charmcraft-yaml-base` key in ``charmcraft.yaml`` defines the operating

> system and version on which a charm can build and run. Supported operating systems

Reviewer: ```suggestion
system and version on which a charm can build and run. Supported OSs
```

vary by both the version of Charmcraft and the version of `Juju`_.

The general syntax for defining a distribution and series is to use
``base: <distribution>@<series>``. For example, a charm that runs on Ubuntu 24.04 LTS may be defined with the following ``base`` key:

.. code-block:: yaml

    base: ubuntu@24.04

Select one build and one target architecture
--------------------------------------------

The following ``charmcraft.yaml`` snippet will define a charm that both builds and

> runs on Ubuntu 24.04 LTS, only on an amd64 architecture:

Reviewer: ```suggestion
runs on Ubuntu 24.04 LTS, only on an AMD64 architecture:
```

There are a few instances elsewhere where it's lowercase.

A charm may require cross-compilation to build. To create a charm for a
different architecture, use the following ``charmcraft.yaml`` snippet:

.. code-block:: yaml

   base: ubuntu@24.04
   platforms:
     riscv64-cross:
       build-on: [amd64]
       build-for: [riscv64]

Building on amd64 will produce one charm that runs on ``riscv64``.

Note that the charm developer must ensure the charm is compatible with the
target architectures. By default, the `charm`_,
:ref:`python <craft_parts_python_plugin>`, :ref:`poetry <craft_parts_poetry_plugin>`,
and :ref:`uv <craft_parts_uv_plugin>` plugins will install wheels for python packages
for the ``build-on`` architecture rather than the ``build-for``. For more information,
see `craft-parts#974`_.

Select multiple OS releases
---------------------------

The resulting ``.charm`` file packed by charmcraft can only run on a single OS release
or ``base``, using ``charmcraft.yaml`` nomenclature. A ``charmcraft.yaml`` can use
multi-base syntax to create a set of charms, each for a different base. To do this, the
base is defined in each platform entry instead of being defined with the
top-level ``base`` and ``build-base`` keys.

> To build a charm for Ubuntu 22.04 and a charm for Ubuntu 24.04, use the

Reviewer: ```suggestion
To build a charm for Ubuntu 22.04 LTS and a charm for Ubuntu 24.04 LTS, use the
```

In the last paragraph, too.

Reviewer 2: IMO this language is cumbersome. Not your fault I know... I just want to complain

following ``charmcraft.yaml`` snippet which uses :ref:`multi-base
notation<reference-platforms-multi-base>`:

.. code-block:: yaml

   platforms:
     ubuntu-22.04-amd64:
       build-on: [ubuntu@22.04:amd64]
       build-for: [ubuntu@22.04:amd64]
     ubuntu-24.04-amd64:
       build-on: [ubuntu@24.04:amd64]
       build-for: [ubuntu@24.04:amd64]

The ``build-on`` and ``build-for`` entries are identical for each platform, so
the :ref:`multi-base shorthand notation
<reference-platforms-multi-base-shorthand>` can be used:

.. code-block:: yaml

   platforms:
     ubuntu@22.04:amd64:
     ubuntu@24.04:amd64:

> With both snippets, building on amd64 will produce two charms, one for
> amd64 systems running Ubuntu 22.04 and one for amd64 systems running
> Ubuntu 24.04.

Reviewer: ```suggestion
With both snippets, building on AMD64 will produce two charms, one for
AMD64 hosts running Ubuntu 22.04 LTS and one for AMD64 hosts running
Ubuntu 24.04 LTS.
```

Since you changed it to "hosts" earlier in the page.

Reviewer 2: I'm distinguishing here the build host from the running system, which is why I kept these as systems.

Reviewer: Non-blocking, but what makes runtime different? Don't charms run on every kind of machine?
